pkg/bloomgateway/bloomgateway.go

```diff
@@ -355,19 +312,250 @@ func (g *Gateway) FilterChunkRefs(ctx context.Context, req *logproto.FilterChunk
 	})
 
 	response := make([]*logproto.GroupedChunkRefs, 0, len(req.Refs))
+	responseCount := 0
 	for {
 		select {
 		case <-ctx.Done():
-			return nil, ctx.Err()
+			return nil, errors.Wrap(ctx.Err(), "waiting for results")
 		case err := <-errCh:
-			return nil, err
+			return nil, errors.Wrap(err, "waiting for results")
 		case res := <-resCh:
-			level.Info(g.logger).Log("msg", "got result", "task", task.ID, "tenant", tenantID, "res", res)
+			responseCount++
+			// log line is helpful for debugging tests
+			// level.Debug(g.logger).Log("msg", "got partial result", "task", task.ID, "tenant", tenantID, "fp", res.Fp, "chunks", res.Chks.Len(), "progress", fmt.Sprintf("%d/%d", responseCount, len(req.Refs)))
+			if res.Chks.Len() > 0 {
+				response = append(response, &logproto.GroupedChunkRefs{
+					Tenant:      tenantID,
+					Fingerprint: uint64(res.Fp),
+					Refs:        convertToShortRefs(res.Chks),
+				})
+			}
 			// wait for all parts of the full response
-			response = append(response, res)
-			if len(response) == len(req.Refs) {
+			if responseCount == len(req.Refs) {
 				return &logproto.FilterChunkRefResponse{ChunkRefs: response}, nil
 			}
 		}
 	}
 }
+
+type workerConfig struct {
+	maxWaitTime time.Duration
+	maxItems    int
+}
+
+type workerMetrics struct {
+	dequeuedTasks      *prometheus.CounterVec
+	dequeueErrors      *prometheus.CounterVec
+	dequeueWaitTime    *prometheus.SummaryVec
+	storeAccessLatency *prometheus.HistogramVec
+}
+
+func newWorkerMetrics(subsystem string, registerer prometheus.Registerer) *workerMetrics {
+	labels := []string{"worker"}
+	return &workerMetrics{
+		dequeuedTasks: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "loki",
+			Subsystem: subsystem,
+			Name:      "dequeued_tasks_total",
+			Help:      "Total amount of tasks that the worker dequeued from the bloom query queue",
+		}, labels),
+		dequeueErrors: promauto.With(registerer).NewCounterVec(prometheus.CounterOpts{
+			Namespace: "loki",
+			Subsystem: subsystem,
+			Name:      "dequeue_errors_total",
+			Help:      "Total amount of failed dequeue operations",
+		}, labels),
+		dequeueWaitTime: promauto.With(registerer).NewSummaryVec(prometheus.SummaryOpts{
+			Namespace: "loki",
+			Subsystem: subsystem,
+			Name:      "dequeue_wait_time",
+			Help:      "Time spent waiting for dequeuing tasks from queue",
+		}, labels),
+		storeAccessLatency: promauto.With(registerer).NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: "loki",
+			Subsystem: subsystem,
+			Name:      "store_latency",
+			Help:      "Latency in seconds of accessing the bloom store component",
+		}, append(labels, "operation")),
+	}
+}
+
+// worker is a datastructure that consumes tasks from the request queue,
+// processes them and returns the result/error back to the response channels of
+// the tasks.
+// It is responsible for multiplexing tasks so they can be processes in a more
+// efficient way.
+type worker struct {
```

Comment:
I think we could move the worker logic to a new sub pkg `pkg/bloomgateway/worker`. And have:
- Logic inside `worker.go`
- Config inside `config.go`
- Metrics inside `metrics.go`

Reply:
IMO that's not necessary. Golang is not like Java where every single class has its own file.
500-ish LOC are not that many for a go file.
The only thing I could imagine is splitting the worker related code into a separate file.

Reply:
I was more concerned about logically splitting files and following a pattern we already follow elsewhere with config and metrics on separate files. Also, I do feel like workers can be logically separated as right now we have workers as goroutines, but maybe in the future we want workers running like we run queriers.

It's just a nit tho, so feel free to disregard this.

Reply:
I did split the worker code in a separate `worker.go` file :)